### PR TITLE
refine rate limiter: per-query token consumption

### PR DESCRIPTION
**Pull request by ryanfkeepers**
Now that the rate limiter is split by service, we can further split by per-query token consumption.  Two primary service cases exist: all exchange queries assume to cost a single token (for now).  Drive-service queries are split between permissions (5), default cost (2), and single-item or delta gets (1).

---

#### Does this PR need a docs update or release note?

- [x] :clock1: Yes, but in a later PR

#### Type of change

- [x] :sunflower: Feature

#### Issue(s)

* #2951

#### Test Plan

- [x] :zap: Unit test
- [x] :green_heart: E2E

### CODE REVIEW

**Review by pandeyabs — COMMENTED** [review]
Hey @ryanfkeepers  could you rebase this on main? Currently this PR is also showing changes from #3357

**Comment by ryanfkeepers** [comment]
@pandeyabs Done!  Thanks for the ping.

File: src/internal/connector/graph/middleware.go [hunk 1]
```diff
@@ -419,13 +424,55 @@ func extractRateLimiterConfig(ctx context.Context) (LimiterCfg, bool) {
 	return lc, ok
 }
 
+type limiterConsumptionKey string
```
Comment (pandeyabs) [inline src/internal/connector/graph/middleware.go:427]: nit: Very subjective, but to me this typedef isn't adding much value.

File: src/internal/connector/graph/middleware.go [hunk 2]
```diff
@@ -419,13 +424,55 @@ func extractRateLimiterConfig(ctx context.Context) (LimiterCfg, bool) {
...
+
+const (
+	limiterConsumptionCtxKey       limiterConsumptionKey = "corsoGraphRateLimiterConsumption"
+	defaultLimiterConsumption                            = 1
+	driveDefaultLimiterConsumption                       = 2
```
Comment (pandeyabs) [inline src/internal/connector/graph/middleware.go:432]: nit: driveLC*? Just to be consistent with other constants defined below.

File: src/internal/connector/graph/middleware.go [hunk 3]
```diff
@@ -419,13 +424,55 @@ func extractRateLimiterConfig(ctx context.Context) (LimiterCfg, bool) {
...
+}
+
+func ctxLimiterConsumption(ctx context.Context, defaultConsumption int) int {
+	l := ctx.Value(limiterConsumptionCtxKey)
+	if l == nil {
+		return defaultConsumption
+	}
+
+	lc, ok := l.(int)
+	if !ok || lc < 1 {
```
Comment (pandeyabs) [inline src/internal/connector/graph/middleware.go:455]: Would be good to add an upper bound check too.

File: src/internal/connector/graph/middleware.go [hunk 4]
```diff
@@ -419,13 +424,55 @@ func extractRateLimiterConfig(ctx context.Context) (LimiterCfg, bool) {
...
+	// limit consumption rate for anything permissions related
+	PermissionsLC = 5
+)
+
+// ConsumeNTokens ensures any calls using this context will consume
+// n rate-limiter tokens.  Default is 1, and this value does not need
+// to be established in the context to consume the default tokens.
+// This should only get used on a per-call basis, to avoid cross-pollination.
+func ConsumeNTokens(ctx context.Context, n int) context.Context {
+	return context.WithValue(ctx, limiterConsumptionCtxKey, n)
```
Comment (pandeyabs) [inline src/internal/connector/graph/middleware.go:445]: Should we validate n here?

**Review by pandeyabs — APPROVED** [review]
Left some comments but overall looks good!

File: src/internal/connector/graph/middleware.go (hunk 1 above)
Comment (ryanfkeepers) [reply src/internal/connector/graph/middleware.go:427]: It's actually not subjective in this case.  Golang context values are a type of `map[any]any` (where, of course, `any` is an alias for the empty interface, `interface{}`).  Using `any` as a key causes the map to store both the value and the type of the key.

This way it's possible to store equally valued strings without collision, but that only works if both values have a unique type.  Go ahead and try to add to a context using `WithValue` and a plain string: you should see a linter error about keying with an untyped string.  

Bit of an ingenious way to ensure that a shared variable with a generic storage interface stays free of common-name collisions.

File: src/internal/connector/graph/middleware.go (hunk 2 above)
Comment (ryanfkeepers) [reply src/internal/connector/graph/middleware.go:432]: sure

File: src/internal/connector/graph/middleware.go (hunk 4 above)
Comment (ryanfkeepers) [reply src/internal/connector/graph/middleware.go:445]: We already have lazy validation, which I think is marginally safer in this case, and doesn't need both ends guarded.

File: src/internal/connector/graph/middleware.go (hunk 3 above)
Comment (ryanfkeepers) [reply src/internal/connector/graph/middleware.go:455]: Upper bounds are naturally checked by the rate limiter.  If a capacity greater than the per-second burst is requested, then the limiter (and thus api call) will error.  Should be clearly identifiable in CI testing, and prevents us from needing to check against N different upper bounds (one for each limiter implementation, should they differ).

File: src/internal/connector/graph/middleware.go (hunk 1 above)
Comment (pandeyabs) [reply src/internal/connector/graph/middleware.go:427]: Thanks! Yeah, this makes total sense.

**Pull request merged**